### ISSUE TITLE
BatchNorm option for exporting to JS by using int32 instead of int64.

### ISSUE THREAD

**Issue by Narsil** [pytorch/pytorch#14807]
## 🚀 Feature
Is it possible to move BatchNorm.num_batches_tracked from `torch.long` to `torch.int` ? Or enable an option for it ?

## Motivation

The motivation is linked to Javascript handling of Integers. Because there is no way to correctly handle Int64 in Javascript, when you export a model to JS (through ONNX and https://github.com/chaosmail/tfjs-onnx for instance) the models are unreadable because of the hard way to interpret the Int64.

## Pitch

The best way would probably be to have an option in the BatchNorm definition.
`BatchNorm2d(....., track_type=torch.int)` as it would not change anything backward.

We could also directly force a `torch.int` type for that value as most people would probably never reach 2Billion batches. But ofc is still is a major change

## Alternatives

Other alternatives would be to treat in the JS libraries these corner cases, which is definitely an option, but they would have to drop some precision no matter what (if you don't want non native types). This could happen before exporting to onnx, or at import time in JS.

Before exporting to onnx seems bad because it would be a very special case for BatchNorm, and some Int64 might have very good reason to be large. The same reason applies to JS import.

Simple export script for context:

````
#!/usr/bin/env python                                                            
import torch                                                                     
from torch import nn                                                             
                                                                                 
                                                                                 
def main():                                                                      
    model = nn.Sequential(                                                       
        nn.Conv2d(3, 8, kernel_size=3),                                         
        nn.BatchNorm2d(8)                                                     
    )                                                                            
    model.eval()                                                                 
                                                                                 
    image = torch.rand(1, 3, 224, 224)                                           
    input_names = ["input"] + [f"tensor_{i}" for i in range(2)]                  
    torch.onnx.export(model, image, "out.onnx", verbose=True, input_names=input_names)
                                                                                 
                                                                                 
if __name__ == "__main__":                                                       
    main()  
```

**Comment by soumith** [comment]
I dont want to set precedent here, where we are re-working our APIs because of a downstream limitation. I think we'll see this be an issue in other corner cases of the APIs as well.

I think it's best to go with your Alternative ideas.

> Other alternatives would be to treat in the JS libraries these corner cases, which is definitely an option, but they would have to drop some precision no matter what (if you don't want non native types). This could happen before exporting to onnx, or at import time in JS.

I think this should be either treated before exporting to ONNX (in which case, the ONNX standard should more strongly type such a field if it exists), or it should be handled at import time in JS.

**Comment by apaszke** [comment]
Yes, this definitely can’t be made an int, because 2^31 training iterations is not an unreasonable number. This should definitely be fixed in JS import. Those 64-bit number are unlikely to be larger than 2^52, and so should fit just fine into the double mantissa (which is the only numeric type in JS, just like in Lua 😊)

**Comment by Narsil** [comment]
Fair enough ! Thank you.

**Comment by waittim** [comment]
Hi Narsil, may I know how do you solve the problem finally? I'm working on the same issue right now....
